test/resolver_test.go

```diff
@@ -0,0 +1,215 @@
+	defer cancel()
+	shorterTimeout := 3 * time.Second
+
+	testMD := metadata.MD{"footest": []string{"bazbar"}}
+	mdOut := metadata.MD{"handler": []string{"value"}}
+
+	var onCommittedCalled bool
+
+	testCases := []struct {
+		name   string
+		md     metadata.MD
+		config *iresolver.RPCConfig
+
+		wantMD       metadata.MD
+		wantDeadline time.Time
+		wantTimeout  time.Duration
+	}{{
+		name:         "basic",
+		md:           testMD,
+		config:       &iresolver.RPCConfig{},
+		wantMD:       testMD,
+		wantDeadline: ctxDeadline,
+	}, {
+		name: "alter MD",
+		md:   testMD,
+		config: &iresolver.RPCConfig{
+			Context: metadata.NewOutgoingContext(ctx, mdOut),
+		},
+		wantMD:       mdOut,
+		wantDeadline: ctxDeadline,
+	}, {
+		name: "alter timeout; remove MD",
+		md:   testMD,
+		config: &iresolver.RPCConfig{
+			Context: longdeadlineCtx, // no metadata
+		},
+		wantMD:       nil,
+		wantDeadline: longCtxDeadline,
+	}, {
+		name:         "nil config",
+		md:           metadata.MD{},
+		config:       nil,
+		wantMD:       nil,
+		wantDeadline: ctxDeadline,
+	}, {
+		name: "alter timeout via method config; remove MD",
+		md:   testMD,
+		config: &iresolver.RPCConfig{
+			MethodConfig: serviceconfig.MethodConfig{
+				Timeout: &shorterTimeout,
+			},
+		},
+		wantMD:      nil,
+		wantTimeout: shorterTimeout,
+	}, {
+		name: "onCommitted callback",
+		md:   testMD,
+		config: &iresolver.RPCConfig{
+			OnCommitted: func() {
+				onCommittedCalled = true
+			},
+		},
+		wantMD:       testMD,
+		wantDeadline: ctxDeadline,
+	}}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			if !sendConfigChan.SendOrFail(tc.config) {
+				t.Fatalf("last config not consumed by config selector")
+			}
+
+			onCommittedCalled = false
+			ctx = metadata.NewOutgoingContext(ctx, tc.md)
+			startTime := time.Now()
+			if _, err := ss.client.EmptyCall(ctx, &testpb.Empty{}); err != nil {
+				t.Fatalf("client.EmptyCall(_, _) = _, %v; want _, nil", err)
+			}
+
+			gotInfoI, ok := gotInfoChan.ReceiveOrFail()
+			if !ok {
+				t.Fatalf("no config selector data")
+			}
+			gotInfo := gotInfoI.(*iresolver.RPCInfo)
+
+			if want := "/grpc.testing.TestService/EmptyCall"; gotInfo.Method != want {
+				t.Errorf("gotInfo.Method = %q; want %q", gotInfo.Method, want)
+			}
+
+			gotContextI, ok := gotContextChan.ReceiveOrFail()
+			if !ok {
+				t.Fatalf("no context received")
+			}
+			gotContext := gotContextI.(context.Context)
+
+			gotMD, _ := metadata.FromOutgoingContext(gotInfo.Context)
+			if diff := cmp.Diff(tc.md, gotMD); diff != "" {
+				t.Errorf("gotInfo.Context contains MD %v; want %v\nDiffs: %v", gotMD, tc.md, diff)
+			}
+
+			gotMD, _ = metadata.FromIncomingContext(gotContext)
+			// Remove entries from gotMD not in tc.wantMD (e.g. authority header).
+			for k, _ := range gotMD {
+				if _, ok := tc.wantMD[k]; !ok {
+					delete(gotMD, k)
+				}
+			}
+			if diff := cmp.Diff(tc.wantMD, gotMD, cmpopts.EquateEmpty()); diff != "" {
```

Comment:
I have never used `cmp.Diff` to compare types. Is the diff output useful?

Reply:
MD is just a map, so it will print the entries that are different and how:
```
                Diffs:   metadata.MD{
                  	"footest": {
                  		"bazbar",
                - 		"barbaz",
                  	},
                - 	"test": {"fake"},
                  }
```